### DISCUSSION TITLE
インデックスを指定して検索させることは可能でしょうか

### DISCUSSION THREAD

**Discussion by hisahisabanbi** [pgroonga/pgroonga#573] [Q&A] [answered]
同じテーブル、カラムに２つのインデックスを作成して、インデックスを指定した検索をしたいですが可能でしょうか。
ドキュメントを見る限りできそうですが、試してみるとできません。

再現環境は下記です。

下記のテーブル、インデックス、データがあります。
pgroongaのバージョンは3.0.7（Supabase環境です）

```sql
-- テーブル
create table documents (
  id bigint primary key generated always as identity,
  content text
);

-- インデックス
CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );

CREATE INDEX idx_documents_content_no_tokenizer ON documents
  USING pgroonga (content)

-- データ
insert into documents (content) values ('焼肉定食');
insert into documents (content) values ('東京タワー');

-- select
SELECT * FROM documents where content &@~ ('ヤキニク', array[1], 'idx_documents_content')::pgroonga_full_text_search_condition; --検索されない
SELECT * FROM documents where content &@~ ('肉', array[1], 'idx_documents_content_no_tokenizer')::pgroonga_full_text_search_condition; --検索される

```

インデックスの作成順序を変えると、データの検索可否が逆になります。後に作成したインデックスが使用される状況です。

**Comment by askdkc** [comment]
@hisahisabanbi さん

Normalizerの指定を入れると上手く動くかと。

こんな感じですね。

```sql
DROP INDEX idx_documents_content;

CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)', 
    normalizers='NormalizerAuto'
  );

SELECT * FROM documents where content &@~ ('ヤキニク', array[1], 'idx_documents_content')::pgroonga_full_text_search_condition;
--  id | content  
-- ----+----------
--   1 | 焼肉定食
-- (1 行)
```

**Comment by hisahisabanbi** [comment]
@askdkc ありがとうございます。
確かに、上記のDROP INDEX、CREATE INDEXすると、
```sql
SELECT * FROM documents where content &@~ ('ヤキニク', array[1], 'idx_documents_content')::pgroonga_full_text_search_condition;
```
は検索できるようになるんだけど、今度は下記が検索されなくなります。
```sql
SELECT * FROM documents where content &@~ ('肉', array[1], 'idx_documents_content_no_tokenizer')::pgroonga_full_text_search_condition;
```
後につくったインデックスが効くようになるようです。

**Reply by askdkc** [reply]
なるほど。

一度インデックスを削除して、再度作り直したら動いたので対応出来たと思ったのですが、単に後から作られたインデックスになって優先されてただけだったのですね。。。

失礼しました。

**Comment by komainu8** [comment]
`NormalizerAuto`はデフォルトで指定されます。（https://pgroonga.github.io/ja/reference/create-index-using-pgroonga.html#customization を参照いただければと思います。）

```sql
CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)', 
    normalizers='NormalizerAuto'
  );
```
と
```sql
CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );
```
は同じです。

`ヤキニク`でヒットしないケースでは、`ヤキニク`で検索するときに`idx_documents_content_no_tokenizer`のインデックスを使って検索しているからです。`ヤキニク`がヒットして`肉`がヒットしないのは、`idx_documents_content`のインデックスを使って検索しているからです。

したがって、`normalizers='NormalizerAuto'`を追加しても結果は変化しないと思われます。（どちらか一方はヒットしない。）

---

やりたいことは以下のようなことでしょうか？

読み仮名で検索するパターンと、漢字で検索するパターンがあって、どちらのパターンでも同じカラムを検索対象として検索したい。
↑の例なら、読み仮名で検索するパターン、漢字で検索するパターンのどちらも`content`カラムを対象に検索したい。

もしそうなら、以下のインデックスを設定するだけでも`ヤキニク`と`焼肉`はヒットします。

```sql
CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );
```

ただ、↑を設定しても`ニク`や`肉`はヒットしません。
TokenMecabは「焼肉定食」の読み仮名を「ヤキニクテイショク -> ヤキニク / テイショク」と分割するので、`ニク`と完全に一致する語がないためです。('ニク'≠'ヤキニク')

`idx_documents_content_no_tokenizer`の方は、トークナイザーが指定されていないので、デフォルトのトークナイザー[`TokenBigram`](http://groonga.org/ja/docs/reference/tokenizers.html#token-bigram) が使われます。
こちらは、'焼肉定食'を'焼肉'/'肉定'/'定食'/'食'と分割するので、`TokenMecab`と挙動が変わり'肉'でもヒットします。（なんで'肉'でもヒットするかは、あまり回答内容と関係がないので割愛します。）

`TokenMecab`を使って読み仮名検索をする場合、どのキーワードでヒットするかは、MeCabが読み仮名をどう分割するかに左右されてしまうので、'焼肉'ではヒットしても'肉'ではヒットしないということもありえます。
ご質問のケースでは、'肉'で'焼肉'もヒットさせたいようなケースでしょうか？

もし、'肉'で'焼肉'もヒットさせたいようなケース（読み仮名検索ではない検索では、`TokenBigram`を使いたいケース）なら、以下の2つが思いつきます。

1. [`pg_hint_plan`](https://pg-hint-plan.readthedocs.io/ja/latest/)を使って検索時に使用するインデックスを強制する。
2. マテリアライズドビューを使う。

1については、リンク先に使い方が記載されているので、そちらを参照いただければと思います。
2については、以下のようにします。

```sql
create table documents (
  id bigint primary key generated always as identity,
  content text
);

-- インデックス
CREATE INDEX idx_documents_content_no_tokenizer ON documents
  USING pgroonga (content);

-- データ
insert into documents (content) values ('焼肉定食');
insert into documents (content) values ('東京タワー');

-- 読み仮名検索用にマテリアライズドビューを作る
CREATE MATERIALIZED view documents_yomigana as select id, content from documents;
--作成したマテリアライズドビューに読み仮名検索用のインデックスを設定する
CREATE INDEX idx_documents_yomigana_content ON documents_yomigana
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );

-- 読み仮名で検索する時は、マテリアライズドビューのカラムを対象に検索する
SELECT * FROM documents_yomigana where content &@~ ('ヤキニク', array[1], 'idx_documents_yomigana_content')::pgroonga_full_text_search_condition;
--  id | content  
-- ----+----------
--   2 | 焼肉定食
-- (1 row)

-- 読み仮名で検索しない時は、通常のテーブルを対象に検索する。
SELECT * FROM documents where content &@~ ('肉', array[1], 'idx_documents_content_no_tokenizer')::pgroonga_full_text_search_condition;
--  id | content  
-- ----+----------
--   2 | 焼肉定食
-- (1 row)
```

マテリアライズドビューを更新する手間ができてしまうのがデメリットですが、使いたいインデックスを使うようにはなります。

私が思いつくのは↑の2案なのですが、誰か他により良い案をお持ちでしたらコメントいただければと思います。

**Comment by komainu8** [comment]
ちなみに`&@~`の右辺に設定できるインデックス名(`... &@~ ('ヤキニク', array[1], 'idx_documents_yomigana_content')::pgroonga_full_text_search_condition`なら、`'idx_documents_yomigana_content'`の部分のことです。)なのですが、これは、インデックスが使用されず、シーケンシャルサーチで検索されたときに参照されるインデックスを指定しています。
そのため、インデックスを使った検索をしているときは参照されません。

**Comment by hisahisabanbi** [comment]
@komainu8 ありがとうございます。
ご推察の通り、読み仮名（「ヤキニク」）でも、TokenBigramによる１文字（「肉」）でも検索させたいのです。

> ちなみに&@~の右辺に設定できるインデックス名(... &@~ ('ヤキニク', array[1], 'idx_documents_yomigana_content')::pgroonga_full_text_search_conditionなら、'idx_documents_yomigana_content'の部分のことです。)なのですが、これは、インデックスが使用されず、シーケンシャルサーチで検索されたときに参照されるインデックスを指定しています。
そのため、インデックスを使った検索をしているときは参照されません。

&@~の右辺に設定するインデックス名は、使用するインデックスを指定できるわけではないんですね。
マテリアライズドビューか[pg_hint_plan](https://pg-hint-plan.readthedocs.io/ja/latest/)の方法があるんですね。
見てみます。

**Reply by komainu8** [reply]
> &@~の右辺に設定するインデックス名は、使用するインデックスを指定できるわけではないんですね。

すみません。。。ちょっと説明が雑だったので、補足します。
`&@~`の右辺に設定するインデックスはシーケンシャルサーチのときには参照されます。

シーケンシャルサーチの時は、当然ですが設定されているインデックスは使用されません。
そうすると、インデックスに設定されているノーマライザーやトークナイザーも使えません。

こうなると、インデックスを使って検索したときとシーケンシャルサーチのときで検索結果が異なってしまうことがあります。
インデックスを使って検索したときは、ノーマライザーやトークナイザーが有効で、シーケンシャルサーチのときはノーマライザーやトークナイザーが無効なためです。

これを防ぐために、シーケンシャルサーチのときでもインデックスのノーマライザーやトークナイザーを参照できるようにするために、 `&@~` の右辺にインデックス名を設定できるようにしています。
こうすることで、シーケンシャルサーチのときでも、`&@~` の右辺に指定したインデックスのノーマライザーやトークナイザーを参照できるようになり、インデックスを使った検索のときとシーケンシャルサーチのときで検索結果が変化しないようになります。

**Comment by kou** [comment, marked as answer]
pg_hint_planを使う方法以外だと、値が変化しない式インデックスを使う方法があります。

PostgreSQLはインデックス作成時の対象カラムの字面と検索時の対象カラムの字面で使えるインデックスを選別しています。`CREATE INDEX ... (content)`が2つだと`SELECT ... content &@~ ...`時にどちらが選ばれるかは実装依存になります。ということで、式インデックスを使ってこの字面を変えてあげるとユーザーがコントロールできます。

↓では`content || ''`と`content || '' || ''`で違う字面にしています。`|| ''`は空文字列を連結する式ですが、空文字列を連結しても元の文字列は変わらないので同じ結果になります。

見た目は悪いのですが、pg_hint_planを使わなくても使いたいインデックスをユーザーがコントロールできます。

```sql
-- テーブル
create table documents (
  id bigint primary key generated always as identity,
  content text
);

-- インデックス
CREATE INDEX idx_documents_content ON documents
  USING pgroonga ((content || ''))
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );

CREATE INDEX idx_documents_content_no_tokenizer ON documents
  USING pgroonga ((content || '' || ''));

-- データ
insert into documents (content) values ('焼肉定食');
insert into documents (content) values ('東京タワー');

-- select
SELECT * FROM documents where (content || '') &@~ ('ヤキニク', array[1], 'idx_documents_content')::pgroonga_full_text_search_condition;
SELECT * FROM documents where (content || '' || '') &@~ ('肉', array[1], 'idx_documents_content_no_tokenizer')::pgroonga_full_text_search_condition;
```

**Reply by hisahisabanbi** [reply]
ありがとうございます。
目的の検索ができそうです。

Supabaseだとpg_hint_planが追加インストールできませんでした。
PostgreSQLのインデックスの選別について知ることができました。ありがとうございました。

**Comment by askdkc** [comment]
`content`をARRAY型でキャストしたタイプに別インデックスでも行けましたね。

```sql
-- テーブル
create table documents (
  id bigint primary key generated always as identity,
  content text
);

-- インデックス
CREATE INDEX idx_documents_content ON documents
  USING pgroonga (content)
  WITH (
    tokenizer='TokenMecab("use_reading", true)'
  );

CREATE INDEX idx_documents_content_no_tokenizer ON documents
  USING pgroonga ((ARRAY[content]::text[]) pgroonga_text_array_full_text_search_ops_v2);

-- データ
insert into documents (content) values ('焼肉定食');
insert into documents (content) values ('東京タワー');

-- select : or で両方検索するようにしておく
select * from documents where content &@~ 'ヤキニク' or ARRAY[content]::text[] &@~ 'ヤキニク';
--  id | content  
-- ----+----------
--   1 | 焼肉定食
-- (1 行)

select * from documents where content &@~ '肉' or ARRAY[content]::text[] &@~ '肉';
--  id | content  
-- ----+----------
--   1 | 焼肉定食
-- (1 行)
```

**Comment by askdkc** [comment]
さらに他のやり方。

```sql
-- 読み仮名用カラム準備
ALTER TABLE documents ADD COLUMN yomigana TEXT;

-- データ追加時に読み仮名を生成してカラムに放り込むようにしておく
CREATE OR REPLACE FUNCTION update_yomigana()
RETURNS TRIGGER AS $$
BEGIN
  NEW.yomigana := (
    SELECT string_agg(t.value, ' ')
    FROM (
      SELECT (unnest(pgroonga_tokenize(NEW.content, 'tokenizer', 'TokenMecab("use_reading", true)')) ->> 'value') AS value
    ) AS t
  );
  RETURN NEW;
END;
$$ LANGUAGE plpgsql;

-- 既存のデータが既にあれば全部の読み仮名をyomiganaカラムに更新
UPDATE documents
SET yomigana = subquery.yomigana
FROM (
  SELECT d.id, string_agg(t.value, ' ') AS yomigana
  FROM documents d,
  LATERAL (
    SELECT (unnest(pgroonga_tokenize(d.content, 'tokenizer', 'TokenMecab("use_reading", true)')) ->> 'value') AS value
  ) AS t
  GROUP BY d.id
) AS subquery
WHERE documents.id = subquery.id;

-- インデックス作成
CREATE INDEX idx_documents_content_no_tokenizer ON documents
  USING pgroonga ((ARRAY[content,yomigana]));

-- 検索
SELECT * FROM documents where ARRAY[content,yomigana] &@~ 'ヤキニク';
--  id | content  |      yomigana       
-- ----+----------+---------------------
--   1 | 焼肉定食 | ヤキニク テイショク
-- (1 行)

SELECT * FROM documents where ARRAY[content,yomigana] &@~ '肉';
--  id | content  |      yomigana       
-- ----+----------+---------------------
--   1 | 焼肉定食 | ヤキニク テイショク
-- (1 行)
```
